### PR TITLE
move topk into oplib

### PR DESCRIPTION
**Pull request by haoyouab**
For those ops that need `TopK` functionality (e.g. `DetectionOutput`) I just re-implement it inside of `detection_output.cpp` in OV bridge. Some other ops that haven't been implemented like `Proposal`, `NonMaxSuppression` in Caffe and MxNet style will need `TopK`, too. Move it into oplib so those ops can call it directly without producing so much redundant code.

### CODE REVIEW

**Review by tzerrell — COMMENTED** [review]
Great! This definitely belongs in the oplib, thanks for moving it there.

File: plaidml/op/op.h [hunk 1]
```diff
@@ -654,6 +660,53 @@ class nms {
...
+class topk {
+ public:
+  explicit topk(edsl::Tensor I, int k) : I_(I), k_(k) {}
+
+  topk& axis(int axis) {
+    axis_ = axis;
+    return *this;
+  }
+
+  topk& mode(edsl::SortDirection mode) {
```
[the remark this answers is not in the document]
Comment (haoyouab) [reply plaidml/op/op.h]: Changed to `sort_direction`.

File: plaidml/op/op.h [hunk 2]
```diff
@@ -654,6 +660,53 @@ class nms {
...
+    auto indices = R[1].as_tensor();
+    return {values, indices};
+  }
+
+ private:
+  edsl::Tensor I_;
+  int k_;
+  int axis_;
+  edsl::SortDirection mode_;
+  TopKSortType sort_type_;
```
[the remark this answers is not in the document]
Comment (haoyouab) [reply plaidml/op/op.h]: Done.

**Review by tzerrell — COMMENTED** [review]
Looks generally good, one comment about style consistency for default values.

File: plaidml/op/op.h [hunk 3]
```diff
@@ -654,6 +660,53 @@ class nms {
...
 };
 
+class topk {
+ public:
+  explicit topk(edsl::Tensor I, int k) : I_(I), k_(k), axis_(-1) {}
+
+  topk& axis(int axis) {
+    axis_ = axis;
+    return *this;
+  }
```
[the remark this answers is not in the document]
Comment (flaub) [reply plaidml/op/op.h:670]: The general rule is, required parameters go into the constructor, optional parameters have builder setters. You can have multiple constructors to represent each configuration of required parameters (like if certain modes have a subset of parameters that must be specified together).

[the remark this answers is not in the document]
Comment (haoyouab) [reply plaidml/op/op.h:670]: > The general rule is, required parameters go into the constructor, optional parameters have builder setters. You can have multiple constructors to represent each configuration of required parameters (like if certain modes have a subset of parameters that must be specified together).

I searched around in `OpLib` and `Edsl` and it seems `axis` is mostly used as an optional parameter. Should we go for option 1 as others do?

[the remark this answers is not in the document]
Comment (tzerrell) [reply plaidml/op/op.h]: Sure, that sounds good to me, treating `axis` as an optional parameter with default value of `-1` makes sense.

**Review by tzerrell — APPROVED** [review]
LGTM, I see you made the necessary changes for the "default axis to -1" approach we discussed in the inline comments.

**Pull request merged**